src/components/editor/SinopiaResourceTemplates.jsx

```diff
@@ -1,95 +1,92 @@
     }
+
+    await this.fetchResourceTemplatesFromGroups()
   }
 
-  fulfillGroupPromise = (promise) => {
-    promise.then(data => {
-      const contains = [].concat(data.response.body.contains)
-      this.setState({groupData: contains})
-    }).catch((error) => {
-      this.setState({message: error})
-    })
-    return promise
+  fetchResourceTemplatesFromGroups = async () => {
+    try {
+      const getGroupsResponse = await getGroups()
+      const contains = [].concat(getGroupsResponse.response.body.contains).filter(Boolean)
+      contains.forEach(async group => {
+        const groupName = resourceToName(group)
+        const listResourcesInGroupResponse = await listResourcesInGroupContainer(groupName)
+        // Short-circuit listing resources in a group if it contains nothing
+        if (listResourcesInGroupResponse.response.body.contains)
+          this.setStateFromServerResponse(groupName, listResourcesInGroupResponse.response.body.contains)
+      })
+    } catch(error) {
+      const errors = [...this.state.errors, error]
+      this.setState({ errors: errors })
+    }
   }
 
-  fulfillGroupData = (data) => {
-    const groupName = resourceToName(data.response.body['@id'])
+  setStateFromServerResponse = (groupName, groupContains) => {
+    // Array-ify the value of `groupContains` as it can be either a string or an array
+    const templateIds = [].concat(groupContains)
 
-    if (data.response.body.contains !== undefined) {
-      const contains = [].concat(data.response.body.contains)
+    templateIds.forEach(async templateId => {
+      const templateName = resourceToName(templateId)
+      const templateResponse = await getResourceTemplate(templateName, groupName)
+      const resourceTemplateBody = templateResponse.response.body
 
-      contains.map(c => {
-        const name = resourceToName(c)
+      const retrievedTemplateObject = {
+        key: resourceTemplateBody.id,
+        name: resourceTemplateBody.resourceLabel,
+        uri: templateId,
+        id: resourceTemplateBody.id,
+        group: groupName
+      }
 
-        this.groupDataPromise(groupName, name).then((data) => {
-          const rt = data.response.body
-          this.setState({tempState: {key: rt.id, name: rt.resourceLabel, uri: c, id: rt.id, group: groupName}})
-          const joined = this.state.templatesForGroup.slice(0)
-          if (!_.find(joined, ['key', this.state.tempState.key])) {
-            joined.push(this.state.tempState)
-          }
-          this.setState({templatesForGroup: joined})
-        }).catch(() => {})
-      })
-    }
-
-  }
+      const templates = [...this.state.resourceTemplates]
 
-  groupDataPromise = (groupName, name) => new Promise(async (resolve) => {
-    await resolve(getResourceTemplate(name, groupName))
-  })
+      // Check if newly retrieved template is already stored in component state. If not, add it.
+      if (!templates.some(template => template.key === retrievedTemplateObject.key))
```

Comment:
👍 Had to lookup `some` array method in the [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some), like it's use here.

Reply:
Alas, I wound up replacing this to fix one of the bugs! Ha.